src/resource_manager/secret.py

```diff
@@ -0,0 +1,167 @@
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+"""gateway-api-integrator secret resource manager."""
+
+import dataclasses
+import logging
+import typing
+
+from cryptography.hazmat.primitives import serialization
+from lightkube import Client
+from lightkube.core.client import LabelSelector
+from lightkube.models.meta_v1 import ObjectMeta
+from lightkube.resources.core_v1 import Secret
+from lightkube.types import PatchType
+from ops.model import Relation
+
+from state.base import State
+
+from .permission import map_k8s_auth_exception
+from .resource_manager import ResourceManager
+
+logger = logging.getLogger(__name__)
+
+
+@dataclasses.dataclass
+class SecretState(State):
+    """A part of charm state with information required to manage secret resource.
+
+    It consistS of 3 components:
+        - SecretResourceInfomation
+        - CharmConfig
+        - TLSInformation
+
+    Attrs:
+        gateway_class: The configured gateway class.
+        external_hostname: The configured gateway hostname.
+        tls_requirer_integration: The integration instance with a TLS provider.
+        tls_certs: A dict of hostname: certificate obtained from the relation.
+        tls_keys: A dict of hostname: private_key stored in juju secrets.
+        secret_resource_name_prefix: Prefix of the secret resource name.
+    """
+
+    gateway_class: str
+    external_hostname: str
+    tls_requirer_integration: Relation
+    tls_certs: dict[str, str]
+    tls_keys: dict[str, dict[str, str]]
```

Comment:
Is there a case where multiple certs and keys are linked to a give `external_hostname` ? If not, shouldn't we directly set the corresponding cert and key ?

Reply:
I left this implementation as-is when bringing it over from the `nginx-ingress-integrator` charm since we're also planning to support multiple hostnames in the future, with this implementation a hostname maps to one cert and one key, so there's no cases where multiple certs are linked to a hostname. I think it's ok to keep it like this, what do you think?

Reply:
Few things to concider:
- The current implementation requests a TLS cert only for the `external_hostname`
- Ingress permits to do `path` routing or `subdomain` routing, which means in the case of `subdomain` routing and multiple relation, we'll have multiple certificates which are subdomains
- TLS relation allows to use SANs (Subject Alternative Names) which allows multiple domains per certificate
- A future `gateway-route` relation will likely allow the requirer to specify which domain they want to use, which means the charm will have to support multiple domains certificate

That said, these's no cases where a single domain has multiple certificates (or just temporarily when a cert is getting renewed).
This data structure associate an `external_hostname` to a list of cert et keys, which doesn't make sense

Reply:
After internal discussion It's decided that we'll rework the attributes in the `SecretResourceDefinition` class in the secret_resource_manager module to only contains a single TLS cert instead of passing the entire dictionary. Since I already have 2 approvals for this PR I'll mark this conversation as resolve and make a follow-up PR to address those changes.